macros/sql/deduplicate.sql

```diff
@@ -16,8 +14,58 @@
                 {%- endif %}
             ) as rn
         from {{ relation if relation_alias is none else relation_alias }} as _inner
-    ) as deduped
-    where deduped.rn = 1
+    )
+
+    select
+        distinct data.*
+    from {{ relation }} as data
+    {#
+    -- Not all DBs will support natural joins but the ones that do include:
+    -- Oracle, MySQL, SQLite, Redshift, Teradata, Materialize, Databricks
+    -- Apache Spark, SingleStore, Vertica
+    -- Those that do not appear to support natural joins include:
+    -- SQLServer, Trino, Presto, Rockset, Athena
+    #}
+    natural join row_numbered
+    where row_numbered.rn = 1
+
+{%- endmacro -%}
+
+{# Redshift should use default instead of Postgres #}
+{% macro redshift__deduplicate(relation, group_by, order_by=none, relation_alias=none) -%}
+
+    {{ return(dbt_utils.default__deduplicate(relation, group_by, order_by=order_by, relation_alias=relation_alias)) }}
+
+{% endmacro %}
+
+{#
+-- Postgres has the `DISTINCT ON` syntax:
+-- https://www.postgresql.org/docs/current/sql-select.html#SQL-DISTINCT
+#}
+{%- macro postgres__deduplicate(relation, group_by, order_by=none, relation_alias=none) -%}
+
+    select
+        distinct on ({{ group_by }}) *
+    from {{ relation if relation_alias is none else relation_alias }}
+    order by {{ group_by }}{{ ',' ~ order_by if order_by is not none else '' }}
+
+{%- endmacro -%}
+
+{#
+-- Snowflake has the `QUALIFY` syntax:
+-- https://docs.snowflake.com/en/sql-reference/constructs/qualify.html
+#}
+{%- macro snowflake__deduplicate(relation, group_by, order_by=none, relation_alias=none) -%}
+
+    select *
+    from {{ relation if relation_alias is none else relation_alias }}
+    qualify
+        row_number() over (
+            partition by {{ group_by }}
+            {% if order_by is not none -%}
```

Comment:
I understand the desire to not make this update a breaking change by requiring a parameter that was previously optional, but if `order_by` is not specified here, the query will fail on Snowflake.